R/derive_advs_params.R

```diff
@@ -0,0 +1,600 @@
+#'   numerator and denominator parameters collected.
+#'
+#'   *Permitted Values*: list of variables created by \code{exprs()}
+#'   e.g. \code{exprs(USUBJID, VISIT)}
+#'
+#' @param get_unit_expr An expression providing the unit of the parameter
+#'
+#'   The result is used to check the units of the input parameters. If the units are not consistent
+#'   within each parameter, an error will be thrown.
+#'
+#'   Additionally, if the input parameters are measured in different units but are mutually
+#'   convertible (e.g., centimeters for one parameter and inches for another), an automatic
+#'   conversion will be performed in order to uniform the values before calculating the ratio.
+#'
+#'   **Note:** Conversion factors comes from unit definitions as per the NCI Thesaurus
+#'
+#'   *Permitted Values:* A variable of the input dataset or a function call
+#'
+#' @inheritParams admiral::derive_param_bmi
+#'
+#' @details
+#' The analysis value of the new parameter is derived as
+#' \deqn{RATIO = \frac{NUMERATOR}{DENOMINATOR}}
+#'
+#'
+#' @return The input dataset with the new parameter added. Note, a variable will only
+#'    be populated in the new parameter rows if it is specified in \code{by_vars}.
+#'
+#' @family internal
+#' @keywords internal
+derive_param_ratio <- function(dataset,
+                               by_vars,
+                               numerator_code,
+                               denominator_code,
+                               set_values_to,
+                               constant_numerator = FALSE,
+                               constant_denominator = FALSE,
+                               filter = NULL,
+                               constant_by_vars = NULL,
+                               get_unit_expr = NULL) {
+  assert_vars(by_vars)
+  assert_data_frame(dataset, required_vars = exprs(!!!by_vars, PARAMCD, AVAL))
+  assert_character_scalar(numerator_code)
+  assert_character_scalar(denominator_code)
+  assert_varval_list(set_values_to, required_elements = "PARAMCD")
+  assert_param_does_not_exist(dataset, set_values_to$PARAMCD)
+  assert_logical_scalar(constant_numerator)
+  assert_logical_scalar(constant_denominator)
+  filter <- assert_filter_cond(enexpr(filter), optional = TRUE)
+  assert_vars(constant_by_vars, optional = TRUE)
+  get_unit_expr <- assert_expr(enexpr(get_unit_expr), optional = TRUE)
+
+  ### Default formula with no units conversion applied ----
+
+  ratio_formula <- expr(
+    !!sym(paste0("AVAL.", numerator_code)) /
+      !!sym(paste0("AVAL.", denominator_code))
+  )
+
+  ### If `get_unit_expr` provided then check units and enable units conversion ----
+
+  if (!missing(get_unit_expr) && !is.null(get_unit_expr)) {
+    param_units <- list()
+
+    # Check if units are the same within each of two parameters
+
+    for (param in c(numerator_code, denominator_code)) {
+      units_found <- dataset %>%
+        mutate(`_unit` = !!get_unit_expr) %>%
```

Comment:
Not aware of this kind of guidelines, it's just based on `admiraldev::assert_unit()`

Reply:
I think this is when you derive a temp var in the dataset that you're going to return. Here it's not applicable because the var created in an intermediate dataset. Although, I agree to replace with `tmp_unit`.

Reply:
Done